docs/enhancements/clusterpool-inventory.md

```diff
@@ -0,0 +1,186 @@
+      message: Currently in use by cluster deployment foo of clusterpool foo-pool
+      reason: ClusterDeploymentCustomizationInUse
+      status: "false"
+      type: Available
+```
+
+if DNS is configured with the name `foo`, ClusterDeploymentCustomization.spec.installConfigPatches content to patch vSphere install config will be as follows
+```yaml
+spec:
+  installConfigPatches:
+    - op: replace
+      path: metadata/name
+      value: foo
+```
+
+When adding a ClusterDeployment, if such an `Inventory` is present, ClusterPool controller will:
+- Load up the inventory list.
+- For each reference it will do a GET to fetch ClusterDeploymentCustomization and check if the `Available` condition is true and `.status.clusterDeploymentRef` is nil.
+- For the first ClusterDeploymentCustomization that is available, it will use the patches in the `spec.installConfigPatchs` field to update the default install config generated by clusterpool. The patches will be applied in the order listed in the inventory. It will also update the status with reference to the cluster deployment that is using it and update the `Available` condition to false.
+- Set the `spec.clusterPoolReference.ClusterDeploymentCustomizationRef`field in the ClusterDeployment with a reference to ClusterDeploymentCustomization CR.
+- Set a Finalizer on the ClusterDeployment that prevents it from getting deleted before releasing the lease on ClusterDeploymentCustomization CR.
+
+Absent an `Inventory`, ClusterPool will continue to use the generated default install config as it does today.
+
+When ClusterDeployment is deleted, ClusterPool controller will: 
+- Find out if ClusterDeployment used an ClusterDeploymentCustomization from `spec.ClusterPoolReference.ClusterDeploymentCustomizationRef`field.
+- If yes, we simply update ClusterDeploymentCustomization, setting the `spec.clusterDeploymentRef` to nil and `Available` to true.  
+
+### How To Use
+For the VSphere case, this allows the administrator to:
+- Preconfigure DNS with following entries (assuming cluster name is `foo`)
+    ```
+    10.0.0.10  api.foo.example.com
+    10.0.0.11  apps.foo.example.com
+    ```
+- Create a ClusterDeploymentCustomization CR to patch `spec.metadata.name` field of the default install config generated by clusterpool controller. Please refer the section above of a sample CR. The content in `spec.installConfigPatches` field should be as follows
+  ```yaml
+  spec:
+  installConfigPatches:
+  - op: replace
+    path: metadata/name
+    value: foo
+  ```
+- Add the name of ClusterDeploymentCustomization CR to`clusterPool.spec.inventory.ClusterDeploymentCustomizations` list. For ClusterDeploymentCustomization with a name `foo-cluster-deployment-customization` the clusterpool should be configured as follows
+    ```yaml
+    spec:
+      inventory:
+        ClusterDeploymentCustomizations:
+          - name: foo-cluster-deployment-customization
+    ```
+  
+### Validation
```

Comment:
What happens in case that none of the entries are valid?
If one valid entry exists then it should update Clusterpool somehow, which field?

Reply:
L131-2 mention using a status condition on the ClusterPool to indicate when any of its ClusterDeploymentCustomizations are missing or invalid. To be more specific (and maybe we should go ahead and add this to the design @akhil-rane) perhaps that condition looks like:

```yaml
    - lastProbeTime: "2020-11-05T14:49:26Z"
      lastTransitionTime: "2020-11-05T14:49:26Z"
      message: "4 out of 7 ClusterDeploymentCustomizations are valid. Missing: foo, bar. Invalid: baz."
      reason: SomeMissingSomeInvalid
      status: "False"
      type: InventoryValid
```

If all entries are missing or invalid, that's just an edge case of "a subset are missing/invalid". Generally:
- The total number of clusters (claimed & unclaimed) will be min(MaxSize, NumValidInventoryEntries)
- The number of unclaimed clusters will be min(Size, MaxSize, ClaimedClusters - NumValidInventoryEntries)
So for the edge case where NumValidInventoryEntries is 0, the pool will simply not be able to produce any clusters.

Reply:
Added the status condition change on L92